modules/tinymce/src/plugins/codesample/main/ts/core/FilterContent.ts

```diff
@@ -46,6 +49,11 @@ const setup = (editor: Editor): void => {
           Prism.get(editor).highlightElement(elm);
           elm.className = Strings.trim(elm.className);
         });
+
+        const endContainer = editor.selection.getRng().endContainer;
+        if (isPre(endContainer)) {
+          editor.selection.select(endContainer);
+        }
```

Comment:
This isn't the correct action for this as it shouldn't highlight the node for this usecase, it should put the selection beside it and it's also going to fail in some cases whereby the selection could be inside another element inside the non-editable pre element.

So the root cause of this bug appears instead to be that it's not setting the contenteditable state when inserting and instead is trying to modify it after. That's never going to end well, so instead what we need to do to fix this is ensure we use a node filter to add the `contenteditable="false"` attribute instead.

Reply:
```typescript
  editor.on('SetContent', () => {
    const dom = editor.dom;
    const unprocessedCodeSamples = Tools.grep(dom.select('pre'), (elm) => {
      return Utils.isCodeSample(elm) && dom.getAttrib(elm, 'data-mce-highlighted') !== 'true';
    });

    if (unprocessedCodeSamples.length) {
      editor.undoManager.transact(() => {
        Tools.each(unprocessedCodeSamples, (elm) => {
          Tools.each(dom.select('br', elm), (elm) => {
            dom.replace(editor.getDoc().createTextNode('\n'), elm);
          });

          elm.innerHTML = dom.encode(elm.textContent ?? '');
          Prism.get(editor).highlightElement(elm);
          dom.setAttrib(elm, 'data-mce-highlighted', true);
          elm.className = Strings.trim(elm.className);
        });
      });
    }
  });

  editor.on('PreInit', () => {
    editor.parser.addNodeFilter('pre', (nodes) => {
      for (let i = 0, l = nodes.length; i < l; i++) {
        const node = nodes[i];
        const isCodeSample = (node.attr('class') ?? '').indexOf('language-') !== -1 ;
        if (isCodeSample) {
          node.attr('contenteditable', 'false');
          node.attr('data-mce-highlighted', 'false');
        }
      }
    });
  });
```
Here is something that I spun up quickly that I believe should resolve this issue. As you can see, in the above we add a node filter to the parser to find codesample pre elements and add the contenteditable state there. This ensures that it'll always be treated as a CEF element and other actions that occur when inserting won't try and modify it either.

Now we still need some way to detect that the codesample needs processing, which is why I've added the `data-mce-highlighted` attribute to track what needs highlighting. Doing this also need `dom.setAttrib(elm, 'data-mce-highlighted', null);` or similar added to the `PreProcess` step above, but I've excluded that just to keep this shorter..